repo/object/object_writer.go

```diff
@@ -289,8 +295,14 @@ func (w *objectWriter) checkpointLocked() (ID, error) {
 	// wait for any in-flight asynchronous writes to finish
 	w.asyncWritesWG.Wait()
 
-	if w.contentWriteError != nil {
-		return EmptyID, w.contentWriteError
+	var err error
+
+	w.contentWriteErrorMutex.Lock()
+	err = w.contentWriteError
```

Comment:
Guarding here the read of `w.contentWriteError` with the mutex because the error can be set outside of the `w.mu` mutex. This allows to reason about 'happened before / after'.

Reply:
@jkowalski maybe I'm missing something here, thoughts?